airflow/models/xcom_arg.py

```diff
@@ -0,0 +1,147 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Any, Dict, List, Union
+
+from airflow.exceptions import AirflowException
+from airflow.models.baseoperator import BaseOperator  # pylint: disable=R0401
+from airflow.models.xcom import XCOM_RETURN_KEY
+
+
+class XComArg:
+    """
+    Class that represents a XCom push from a previous operator.
+    Defaults to "return_value" as only key.
+
+    Current implementations supports
+        xcomarg >> op
+        xcomarg << op
+        op >> xcomarg   (by BaseOperator code)
+        op << xcomarg   (by BaseOperator code)
+
+    **Example**: The moment you get a result from any operator (functional or regular) you can ::
+
+        any_op = AnyOperator()
+        xcomarg = XComArg(any_op) OR xcomarg = any_op.output
+        my_op = MyOperator()
+        my_op >> xcomarg
+
+    This object can be used in legacy Operators via Jinja.
+
+    **Example**: You can make this result to be part of any generated string ::
+
+        any_op = AnyOperator()
+        xcomarg = any_op.output
+        op1 = MyOperator(my_text_message=f"the value is {xcomarg}")
+        op2 = MyOperator(my_text_message=f"the value is {xcomarg['topic']}")
+
+    :param operator: operator to which the XComArg belongs to
+    :type operator: airflow.models.baseoperator.BaseOperator
+    :param key: key value which is used for xcom_pull (key in the XCom table)
+    :type key: str
+    """
+
+    def __init__(self, operator: BaseOperator, key: str = XCOM_RETURN_KEY):
+        self._operator = operator
+        self._key = key
+
+    def __eq__(self, other):
+        return (self.operator == other.operator
+                and self.key == other.key)
+
+    def __lshift__(self, other):
+        """
+        Implements xcomresult << op
+        """
+        self.set_upstream(other)
+        return self
+
+    def __rshift__(self, other):
+        """
+        Implements xcomresult >> op
+        """
+        self.set_downstream(other)
+        return self
+
+    def __getitem__(self, item):
+        """
+        Implements xcomresult['some_result_key']
+        """
+        return XComArg(operator=self.operator, key=item)
+
+    def __str__(self):
+        """
+        Backward compatibility for old-style jinja used in Airflow Operators
+
+        **Example**: to use XArg at BashOperator::
+
+            BashOperator(cmd=f"... { xcomarg } ...")
+
+        :return:
+        """
+        xcom_pull_kwargs = [f"task_ids='{self.operator.task_id}'",
+                            f"dag_id='{self.operator.dag.dag_id}'",
+                            ]
+        if self.key is not None:
+            xcom_pull_kwargs.append(f"key='{self.key}'")
```

Comment:
Is it possible that `self.key` will be None? The default value is `XCOM_RETURN_KEY` and as far as I understand the key has to be provided. If it's so, we should have a check for that in `__init__`.

Reply:
Key is none is only possible if the base class is overridden (MyXComArg inherits XComArg but doesn't provide a key at all, for example).
I think this is good functionality in case your xcom system doesn't have a key necessarily

Reply:
Wait, didn't we make XCom overwritable instead? I guess you can overwrite it and set it on the `output` property, but not sure if I understand the use for that?

Reply:
Not sure exactly what you meant cas, can you elaborate on "overwritable instead"?

Reply:
How is XComArg overridden? XComArg is not easy to be overwritten currently. Can you show an example use case where XComArg is overridden by a subclass?

Reply:
XCom system itself supports `key=None`, so our goal with this was to allow that kind of usage, so if XCom backend uses `key=None` we already have support for it in XComArg.

Reply:
sgtm. Maybe we should still print that key value is None. But not blocking for me.